### PR TITLE
Screen to compare the multiple studies

### PR DESCRIPTION
**Pull request by keisuke-umezawa**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## What does this implement/fix? Explain your changes.

Screen to compare the multiple studies includeing

- Link from the new study list screen to this screen
- Screen for the multiple studies
- Show the following graphs
    - history
    - edf

PR-280

### CODE REVIEW

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/CompareStudies.tsx:102]: How about graying out the list items when `n_objectives` does not match?

Screenshot 2023-02-07 18 16 31

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/CompareStudies.tsx:70]: Like `TrialList` component, how about selecting the first item in Study list by default?

File: optuna_dashboard/ts/trialFilter.ts
Comment (c-bata) [inline optuna_dashboard/ts/trialFilter.ts:119]: Could you please `useFilteredTrials` to improve the re-rendering performance?

File: optuna_dashboard/ts/components/GraphHistories.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphHistories.tsx:36]: Do you think we can unify this component with GraphHistory?

File: optuna_dashboard/ts/components/GraphEdfs.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphEdfs.tsx:24]: Do you think we can unify this component with GraphEdf?

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [reply optuna_dashboard/ts/components/CompareStudies.tsx:78]: How about also adding a breadcrumb entry when navigation to the comparison page?

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/CompareStudies.tsx:75]: I specified `numbers=0` here. What do you think about this solution?
https://github.com/optuna/optuna-dashboard/pull/380/files#diff-3b4445bde53115493e67a33733abac82ab9c1e1c7f3fd447091b80971898d697R190

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/CompareStudies.tsx:128]: As we discussed it, we will show an error message if a user tries to add a study which has different objective numbers.

File: optuna_dashboard/ts/components/GraphEdfs.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphEdfs.tsx]: Unified.

File: optuna_dashboard/ts/components/GraphHistories.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphHistories.tsx]: Unified.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/CompareStudies.tsx]: Fixed and added it.

**Comment by keisuke-umezawa** [comment]
@c-bata @hvy 
Thank you for your review! I followed your comment. PTAL.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [reply optuna_dashboard/ts/components/CompareStudies.tsx]: Thanks, works great!

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [reply optuna_dashboard/ts/components/CompareStudies.tsx:128]: Seems to work, thanks.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [inline optuna_dashboard/ts/components/CompareStudies.tsx:239]: Does this switch do anything with the current plots? For instance, for the optimization history plot, pruned trials are never plotted (since they don't have necessarily have any associated values). Not a strong opinion but we could remove/hide the toggle for now.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [inline optuna_dashboard/ts/components/CompareStudies.tsx:153]: Nit.
```suggestion
                                "You can only compare studies with that has the same number of objectives.",
```

**Review by hvy — COMMENTED** [review]
Seems great trying it out, basically LGTM. I hope @c-bata can take a more in depth look at the code.

Some minor thoughts, I think most can be ignored for the purpose of this PR.

- Can we in the future add a tip to the user, that describes how to select multiple studies? Shift-clicking is the only way in my understanding. Might not be obvious, or might be given that it's the same in the trial list view.
- Quite a lot of colors in the optimization history graph when comparing two or more studies. Would be good if we can reduce?
- We limit comparing studies with different number of directions, but allow comparing studies with different directions (minimize vs maximize). Not sure if this is very intuitive. 
- Should `/compare-studies` be followed by a `/beta`, as it's part of the new experimental UI?
- It might be useful if we could jump from the comparison page to study details pages in the future.

File: optuna_dashboard/ts/trialFilter.ts
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/trialFilter.ts]: I created the version of useFilteredTrials for multiple studies.

File: optuna_dashboard/ts/components/GraphEdf.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphEdf.tsx:79]: Sorry, I just copied it from History plot.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/CompareStudies.tsx:239]: Ok. That could be. Wait for @c-bata review.

**Comment by keisuke-umezawa** [comment]
> Can we in the future add a tip to the user, that describes how to select multiple studies? Shift-clicking is the only way in my understanding. Might not be obvious, or might be given that it's the same in the trial list view.

Yes. The explanation can be added together with TrialList somewhere. Maybe, it would be better to explain it in Tutorial?

> Quite a lot of colors in the optimization history graph when comparing two or more studies. Would be good if we can reduce?

Yes, we can. But, the current implementation is aligned with optuna implementation.

> We limit comparing studies with different number of directions, but allow comparing studies with different directions (minimize vs maximize). Not sure if this is very intuitive.

We can add the condition of directions when selecting studies. Should I do it?

> Should /compare-studies be followed by a /beta, as it's part of the new experimental UI?

Not sure. For example, other beta pages (e.g. /dashboard/studies/0/analytics) are not always with `beta`.

> It might be useful if we could jump from the comparison page to study details pages in the future.

Yes, I think so too. I can create an issue for it.

File: optuna_dashboard/ts/components/TrialList.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/TrialList.tsx:750]: Could you revert this change?

File: optuna_dashboard/ts/components/GraphEdf.tsx
Comment (hvy) [reply optuna_dashboard/ts/components/GraphEdf.tsx:79]: Noticed you changed to `e`, thanks!

**Comment by hvy** [comment]
>> Can we in the future add a tip to the user, that describes how to select multiple studies? Shift-clicking is the only way in my understanding. Might not be obvious, or might be given that it's the same in the trial list view.

> Yes. The explanation can be added together with TrialList somewhere. Maybe, it would be better to explain it in Tutorial?

It's outside the scope of this PR but some additional graphical guidance or a text tip would be more user friendly in my opinion. Describing it in the tutorial sounds like something we could do as well, though.

>> Quite a lot of colors in the optimization history graph when comparing two or more studies. Would be good if we can reduce?

> Yes, we can. But, the current implementation is aligned with optuna implementation.

Understood, thanks.

>> We limit comparing studies with different number of directions, but allow comparing studies with different directions (minimize vs maximize). Not sure if this is very intuitive.

> We can add the condition of directions when selecting studies. Should I do it?

If we are to limit studies to the number of directions, I'd lean towards limiting the directions as well, to be strict about comparable studies, i.e. those that optimize the exact same objective(s). What do you think? Cc. @c-bata 

>> Should /compare-studies be followed by a /beta, as it's part of the new experimental UI?

> Not sure. For example, other beta pages (e.g. /dashboard/studies/0/analytics) are not always with `beta`.

Ah, I see. Do you know about any policies regarding this @c-bata ?

>> It might be useful if we could jump from the comparison page to study details pages in the future.

> Yes, I think so too. I can create an issue for it.

Sounds good! 👍

File: optuna_dashboard/ts/components/TrialList.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/TrialList.tsx]: Reverted.

File: optuna_dashboard/ts/trialFilter.ts
Comment (c-bata) [inline optuna_dashboard/ts/trialFilter.ts:143]: Do you think we can refactor these lines like below?

```suggestion
  studies.map(s => useFilteredTrials(s, targets, filterComplete, filterPruned))
```

**Review by c-bata — CHANGES_REQUESTED** [review]
dos-http-requests

There seems to be a bug that issues a large number of HTTP requests when moving to the Study comparison page. Could you fix it?

File: optuna_dashboard/ts/trialFilter.ts
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/trialFilter.ts]: If the number of studies becomes different, we call a different number of `useMemo`, and it outputs the following warnings and errors.

```
react_devtools_backend.js:2655 Warning: React has detected a change in the order of Hooks called by GraphHistory. This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://reactjs.org/link/rules-of-hooks
```

I created `getFilteredTrials` to unify the impementation in 
[c7c227a](https://github.com/optuna/optuna-dashboard/pull/380/commits/c7c227a24a7dd40c6c00f8ae62bdb3675b4a5f34). How about this solution?

**Comment by keisuke-umezawa** [comment]
@c-bata 
By [058f75c](https://github.com/optuna/optuna-dashboard/pull/380/commits/058f75ce7cce0ab8029647bec36169b072bdc7db), I fixed the bug.

**Comment by keisuke-umezawa** [comment]
@hvy 
Thank you for the comment.

> > > We limit comparing studies with different number of directions, but allow comparing studies with different directions (minimize vs maximize). Not sure if this is very intuitive.
> > 
> > We can add the condition of directions when selecting studies. Should I do it?
> 
> If we are to limit studies to the number of directions, I'd lean towards limiting the directions as well, to be strict about comparable studies, i.e. those that optimize the exact same objective(s). What do you think? Cc. @c-bata

This part is addressed by [8744a84](https://github.com/optuna/optuna-dashboard/pull/380/commits/8744a848b8de72b47f818d5c47db335b377b1efd)

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [inline optuna_dashboard/ts/components/CompareStudies.tsx:164]: Nit.
```suggestion
                                "You can only compare studies that has the same number of objectives.",
```

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [inline optuna_dashboard/ts/components/CompareStudies.tsx:178]: Ditto.
```suggestion
                                "You can only compare studies that has the same directions.",
```

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (hvy) [inline optuna_dashboard/ts/components/CompareStudies.tsx:146]: Not a strong opinion but `selectedIds` might be clearer given the logic.
```suggestion
                          const selectedIds = selected.map(
```

**Comment by hvy** [comment]
> This part is addressed by [8744a84](https://github.com/optuna/optuna-dashboard/pull/380/commits/8744a848b8de72b47f818d5c47db335b377b1efd)

Thanks! 
Currently, we only display the number of objectives, and not their directions, in the leftmost list, which could perhaps make the UI somewhat difficult to navigate through. But I don't think you have to address that in this first PR.

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (hvy) [inline optuna_dashboard/ts/components/GraphHistory.tsx:317]: 📝 A minor note but we now always append the name of the study, even when there's only a single study.

**Comment by keisuke-umezawa** [comment]
@hvy Thank you for reviewing! I fixed nits part.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/CompareStudies.tsx]: I followed TrialList implementation.

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphHistory.tsx:31]: It looks that only `historyPlotInfos[0].objective_names` is used, do you think we need to add `objective_names` here?

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphHistory.tsx:168]: Could you avoid using `!` to unwrap `StudyDetail | null`? It may crash the entire application.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (c-bata) [reply optuna_dashboard/ts/components/CompareStudies.tsx:265]: > Does this switch do anything with the current plots?

Yes. If `includePrunedTrial` is true, the latest intermediate value is used for the pruned trials like below:

include-pruned

So I think we can leave this toggle.

**Review by c-bata — COMMENTED** [review]
Thank you for the update. I left some review comments.

**Review by hvy — COMMENTED** [review]
Thanks for the fixes.
I think it LGTM after @c-bata -san's comments have been addressed.

File: optuna_dashboard/ts/components/GraphEdf.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphEdf.tsx:79]: Sorry, I might wrongly change it during rebasing it. I again fixed it.

File: optuna_dashboard/ts/components/StudyHistory.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/StudyHistory.tsx:98]: Added the condition of `studyDetail !== null`

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphHistory.tsx:31]: In this part, objective_names of all studies are used.

https://github.com/optuna/optuna-dashboard/blob/a9517b4cdc5d371d5bcf6c2bf0b5dcc2818c9a33/optuna_dashboard/ts/components/GraphHistory.tsx#L265

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphHistory.tsx:168]: Sorry if I understand your question wrongly, but this `!` is for logical operator, not for `non-null assertion operator`.

**Comment by keisuke-umezawa** [comment]
@c-bata 
Thank you for reviewing it!
Basically, I followed your comments.

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (c-bata) [reply optuna_dashboard/ts/components/GraphHistory.tsx:168]: Oh, sorry. You are correct. I misunderstood the code. Let me resolve this.

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (c-bata) [reply optuna_dashboard/ts/components/GraphHistory.tsx:31]: I see. That makes sense!

Comment (c-bata) [reply optuna_dashboard/ts/components/GraphHistory.tsx:31]: What do you think about directions?

Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/GraphHistory.tsx:31]: `directions` are also used to make a best line for each study now.

Comment (c-bata) [reply optuna_dashboard/ts/components/GraphHistory.tsx:31]: I see. Let me resolve this conversation 👍

File: optuna_dashboard/ts/components/StudyDetailBeta.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/StudyDetailBeta.tsx:105]: ditto.
https://github.com/optuna/optuna-dashboard/pull/380/files#r1118855931

File: optuna_dashboard/ts/components/StudyDetail.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/StudyDetail.tsx:169]: How about passing an empty list and display an empty graph? It can increase the CLS score.

File: optuna_dashboard/ts/components/StudyDetail.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/StudyDetail.tsx:201]: ditto.

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphHistory.tsx:48]: [Just FYI] After applied my following suggestions, you need to update this line.

https://github.com/optuna/optuna-dashboard/pull/380/files#r1118855931

**Review by c-bata — CHANGES_REQUESTED** [review]
Sorry for the delay 🙇  I left some suggestions.

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphHistory.tsx:49]: `Target` must be prepared by passing all studies, not only the first study. Otherwise, we cannot display the user attributes correctly. Could you also update the `useObjectiveAndUserAttrTargets`?

File: optuna_dashboard/ts/components/GraphHistory.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/GraphHistory.tsx:118]: Doesn't this panic when an empty study list is given?

**Review by c-bata — CHANGES_REQUESTED** [review]

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/CompareStudies.tsx:164]: Fixed.

File: optuna_dashboard/ts/components/CompareStudies.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/CompareStudies.tsx:178]: Fixed.

File: optuna_dashboard/ts/trialFilter.ts
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/trialFilter.ts:95]: Fixed.

**Comment by keisuke-umezawa** [comment]
@c-bata 
As we discussed it in the meeting, I separated the implementation of GraphEdf and GraphHistory, and reverted them to the previous implementations. In the follow-up PR, I will merge those implementation later.

**Review by c-bata — APPROVED** [review]
LGTM with nits.

**Comment by keisuke-umezawa** [comment]
@c-bata I followed your review comments. Could you check it again?

**Pull request merged**
